python/ray/ml/checkpoint.py

```diff
@@ -1,35 +1,126 @@
-import abc
-from typing import Dict, Optional
+import shutil
+import tarfile
+import tempfile
+
+import os
+from typing import Optional
 
 import ray
+from ray import cloudpickle as pickle
+from ray.util.annotations import DeveloperAPI
+from ray.util.ml_utils.cloud import (
+    upload_to_bucket,
+    is_cloud_target,
+    download_from_bucket,
+)
+
+
+DICT_CHECKPOINT_FILE_NAME = "dict_checkpoint.pkl"
+FS_CHECKPOINT_KEY = "fs_checkpoint"
+BYTES_DATA_KEY = "bytes_data"
+
+
+class Checkpoint:
+    """Ray ML Checkpoint.
+
+    This implementation provides interfaces to translate between
+    different checkpoint storage locations: Local FS storage, remote
+    node FS storage, data object, and cloud storage location.
 
+    The constructor is a private API, instead the ``from_`` methods should
+    be used to create checkpoint objects
+    (e.g. ``Checkpoint.from_directory()``).
```

Comment:
Can you add examples and use cases of what guarantees we provide? For example, can any Checkpoint be round-tripped from bytes->dict->path etc? Are any of these conversions inefficient / not typically useful?

Can also short descriptions of how pytorch / mlflow can work with checkpoints to motivate the interface.

Reply:
Added more docs and two examples